lib/Optimizer/Transforms/StatePreparation.cpp

```diff
@@ -0,0 +1,209 @@
+
+#include "PassDetails.h"
+#include "StateDecomposer.h"
+#include "cudaq/Optimizer/Builder/Runtime.h"
+#include "cudaq/Optimizer/CodeGen/QIRFunctionNames.h"
+#include "cudaq/Optimizer/Dialect/CC/CCOps.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeOps.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeTypes.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Conversion/LLVMCommon/TypeConverter.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Complex/IR/Complex.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Target/LLVMIR/TypeToLLVM.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/Passes.h"
+#include "mlir/Transforms/RegionUtils.h"
+#include <span>
+
+namespace cudaq::opt {
+#define GEN_PASS_DEF_STATEPREPARATION
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+#define DEBUG_TYPE "state-preparation"
+
+using namespace mlir;
+
+/// Replace a qubit initialization from vectors with quantum gates.
+/// For example:
+///
+///
+/// Before StatePreparation (state-prep):
+///
+/// module {
+///   func.func @foo() attributes {
+///     %0 = cc.address_of @foo.rodata_0 : !cc.ptr<!cc.array<complex<f32> x 4>>
+///     %1 = quake.alloca !quake.veq<2>
+///     %2 = quake.init_state %1, %0 : (!quake.veq<2>,
+///       !cc.ptr<!cc.array<complex<f32> x 4>>) -> !quake.veq<2> return
+///  }
+///  cc.global constant @foo.rodata_0 (dense<[(0.707106769,0.000000e+00),
+///      (0.707106769,0.000000e+00), (0.000000e+00,0.000000e+00),
+///      (0.000000e+00,0.000000e+00)]> : tensor<4xcomplex<f32>>) :
+///    !cc.array<complex<f32> x 4>
+/// }
+///
+/// After StatePreparation (state-prep):
+///
+/// module {
+///   func.func @foo() attributes {
+///     %0 = quake.alloca !quake.veq<2>
+///     %c1_i64 = arith.constant 1 : i64
+///     %1 = quake.extract_ref %0[%c1_i64] : (!quake.veq<2>, i64) -> !quake.ref
+///     %cst = arith.constant 0.000000e+00 : f64
+///     quake.ry (%cst) %1 : (f64, !quake.ref) -> ()
+///     %c0_i64 = arith.constant 0 : i64
+///     %2 = quake.extract_ref %0[%c0_i64] : (!quake.veq<2>, i64) -> !quake.ref
+///     %cst_0 = arith.constant 0.78539816339744839 : f64
+///     quake.ry (%cst_0) %2 : (f64, !quake.ref) -> ()
+///     quake.x [%1] %2 : (!quake.ref, !quake.ref) -> ()
+///     %cst_1 = arith.constant 0.78539816339744839 : f64
+///     quake.ry (%cst_1) %2 : (f64, !quake.ref) -> ()
+///     quake.x [%1] %2 : (!quake.ref, !quake.ref) -> ()
+///     return
+///   }
+/// }
+
+namespace {
+
+std::vector<std::complex<double>>
+readConstantArray(mlir::OpBuilder &builder, cudaq::cc::GlobalOp &global) {
+  std::vector<std::complex<double>> result{};
+
+  auto attr = global.getValue();
+  auto type = global.getType().getElementType();
+
+  auto arrayTy = dyn_cast<cudaq::cc::ArrayType>(type);
+  assert(arrayTy);
+  assert(attr.has_value());
```

Comment:
nit: might use `&& "must be cc.array"` as more of a description.

I assume this is purely for debug purposes. If so, these can be removed and the `dyn_cast` above converted to `cast`. It gives the same level of assurance, etc.

Reply:
updated